python/kotekan/shared_memory_buffer.py

```diff
@@ -0,0 +1,539 @@
+"""Python interface for the kotekan ring buffer using a shared memory region."""
+
+import logging
+import mmap
+import numpy as np
+import os
+import posix_ipc
+import struct
+
+from kotekan.visbuffer import VisRaw
+
+FMT_UINT64_T = "Q"
+FMT_INT64_T = "q"
+SIZE_UINT64_T = 8
+
+logger = logging.getLogger(__name__)
+
+
+class SharedMemoryError(Exception):
+    """An error has occurred when trying to access the shared memory."""
+
+    pass
+
+
+class SharedMemoryReaderUsageError(Exception):
+    """There was an error in the usage of SharedMemoryReader."""
+
+    pass
+
+
+class NoNewDataError(Exception):
+    """The data in the shared memory has not changed since the last time checked."""
+
+    pass
+
+
+class SharedMemoryReader:
+    """
+    Reader for the shared memory region.
+
+    Parameters
+    ----------
+    semaphore_name : str
+        Full path and file name of semaphore.
+    shared_memory_name : int
+        Full path and file name of shared memory.
+    buffer_size : int
+        Number of time slots to keep in local copy of data. Increase if large amount of data is to
+        be read repeatedly. Decrease for less memory usage.
+    """
+
+    num_structural_params = 6
+    size_structural_data = SIZE_UINT64_T * num_structural_params
+    size_access_record_entry = SIZE_UINT64_T
+    valid_field_padding = 3
+    size_valid_field = 1
+    invalid_value = -1
+
+    def __init__(self, semaphore_name, shared_memory_name, buffer_size):
+        if buffer_size < 1:
+            raise NotImplementedError("Buffer size of 0 is not implemented yet.")
+        self.buffer_size = buffer_size
+
+        # maps from access record timestamp to index (0..buffer_size)
+        self._time_index_map = {}
+
+        self.semaphore = posix_ipc.Semaphore(semaphore_name)
+
+        self.shared_mem_name = shared_memory_name
+        shared_mem = posix_ipc.SharedMemory(shared_memory_name)
+
+        # 0 means entire file
+        self.shared_mem = mmap.mmap(shared_mem.fd, 0, prot=mmap.PROT_READ)
+
+        (
+            self.num_writes,
+            self.num_time,
+            self.num_freq,
+            self.size_frame,
+            self.size_frame_meta,
+            self.size_frame_data,
+        ) = self._read_structural_data()
+
+        if self.num_writes == self.invalid_value:
+            raise SharedMemoryError(
+                "The shared memory referenced by '{}' was marked as invalid by the writer.".format(
+                    self.shared_mem_name
+                )
+            )
+
+        self.len_data = self.num_freq * self.num_time
+
+        self.size_access_record = self.len_data * self.size_access_record_entry
+        self.pos_access_record = self.size_structural_data
+
+        self.size_data = self.len_data * self.size_frame
+        self.pos_data = self.pos_access_record + self.size_access_record
+
+        self._initial_validation()
+
+        os.close(shared_mem.fd)
+
+        self._data = np.ndarray(
```

Comment:
Given that you don't have the vis structural parameters here (i.e. num_inputs) etc, you can't create a full view, but you can create a datatype which exposes three parts: the explicit valid field, and then you can just have `np.void` regions for the metadata and data sections. That will allow you to easily manipulate the valid field as suggested below.